paimon-spark/paimon-spark-3.3/src/main/scala/org/apache/paimon/spark/catalyst/analysis/Spark3ResolutionRules.scala

```diff
@@ -0,0 +1,245 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.spark.catalyst.analysis
+
+import org.apache.paimon.CoreOptions
+import org.apache.paimon.CoreOptions.MergeEngine
+import org.apache.paimon.spark.SparkTable
+import org.apache.paimon.spark.catalyst.Compatibility
+import org.apache.paimon.spark.commands.{PaimonShowTablePartitionCommand, PaimonShowTablesExtendedCommand}
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.analysis.{NamedRelation, PartitionSpec, ResolvedNamespace, UnresolvedPartitionSpec}
+import org.apache.spark.sql.catalyst.expressions.{Alias, ArrayTransform, Attribute, CreateStruct, Expression, GetArrayItem, GetStructField, If, IsNull, LambdaFunction, Literal, NamedExpression, NamedLambdaVariable}
+import org.apache.spark.sql.catalyst.plans.logical.{InsertIntoStatement, LogicalPlan, Project, ShowTableExtended}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.util.CharVarcharUtils
+import org.apache.spark.sql.connector.catalog.Identifier
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+import org.apache.spark.sql.types.{ArrayType, DataType, IntegerType, Metadata, StructField, StructType}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+case class Spark3ResolutionRules(session: SparkSession)
+  extends Rule[LogicalPlan]
+  with SQLConfHelper {
+
+  import org.apache.spark.sql.connector.catalog.PaimonCatalogImplicits._
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsDown {
+    case ShowTableExtended(
+          ResolvedNamespace(catalog, ns),
+          pattern,
+          partitionSpec @ (None | Some(UnresolvedPartitionSpec(_, _))),
+          output) =>
+      partitionSpec
+        .map {
+          spec: PartitionSpec =>
+            val table = Identifier.of(ns.toArray, pattern)
+            val resolvedSpec =
+              PaimonResolvePartitionSpec.resolve(catalog.asTableCatalog, table, spec)
+            PaimonShowTablePartitionCommand(output, catalog.asTableCatalog, table, resolvedSpec)
+        }
+        .getOrElse {
+          PaimonShowTablesExtendedCommand(catalog.asTableCatalog, ns, pattern, output)
+        }
+
+    case insertPlan @ InsertIntoStatement(
```

Comment:
Interesting function, so that we don't need a higher version of Spark and we can only specify some columns to update.

---

[SPARK-38795][SQL] Support INSERT INTO user specified column lists with DEFAULT values

https://issues.apache.org/jira/browse/SPARK-38795